Final draft of Spin watch docs - ready for approval (#601)
fermyon/developer: content/spin/running-apps.md

Some trigger types support additional `spin up` flags.  For example, HTTP applications can have a `--listen` flag to specify an address and port to listen on.  See the [HTTP trigger](http-trigger) and [Redis trigger](redis-trigger) pages for more details.

> ## The Spin Watch Feature

Reviewer: Suggest "The `watch` Command" or "Monitoring Applications for Changes"

Author: Thanks, went with "Monitoring Applications for Changes"

> Spin's `watch` feature can rebuild and restart Spin applications whenever files change. You can use the `spin watch` [command](https://developer.fermyon.com/common/cli-reference#watch) in place of the `spin build` and `spin up` commands, to build, run and then keep your Spin application running without manual intervention.

Reviewer: ```suggestion
Spin's `watch` command rebuilds and restarts Spin applications whenever files change. You can use the `spin watch` [command](https://developer.fermyon.com/common/cli-reference#watch) in place of the `spin build` and `spin up` commands, to build, run and then keep your Spin application running without manual intervention while staying on the latest code and files.
```

Author: Updated in new push.

> > The initial Spin `watch` command accepts valid Spin [up](https://developer.fermyon.com/common/cli-reference#up) options and passes them through to `spin up` for you, when running/rerunning the Spin application.

Reviewer: ```suggestion
> The `watch` command accepts valid Spin [up](https://developer.fermyon.com/common/cli-reference#up) options and passes them through to `spin up` for you when running/rerunning the Spin application.
```

Author: Updated in next push.

> > The initial Spin `watch` command accepts valid Spin [up](https://developer.fermyon.com/common/cli-reference#up) options and passes them through to `spin up` for you, when running/rerunning the Spin application.

Reviewer: ```suggestion
> The `watch` command accepts valid Spin [up](https://developer.fermyon.com/common/cli-reference#up) options and passes them through to `spin up` for you when running/rerunning the Spin application.
```

> By default, Spin watch monitors the application manifest (`spin.toml` file) and will rebuild the running application if the manifest changes. Spin watch also monitors any files specified in the `component.build.watch` section of the `spin.toml` file. For example, the following configuration (belonging to a Spin `http-rust` application) will be rebuilt via `cargo build --target wasm32-wasi --release` and then rerun using the initial `spin up` command whenever changes occur in either Rust (`.rs`) source files or the `Cargo.toml` file. Glob patterns (i.e. the use of the `*` wildcard) are accepted, as shown below:

Reviewer: ```suggestion
By default, Spin watch monitors:

* The application manifest (`spin.toml` file)
* Any files specified in the `component.build.watch` section of the `spin.toml` file
* Any files specified in the `component.files` section of the `spin.toml` file
* The files specified in the `component.source` section of the `spin.toml` file

If any of these change, Spin will rebuild the application if necessary, then restart the application with the new files.

For example, the following configuration (belonging to a Spin `http-rust` application) will be rebuilt via `cargo build --target wasm32-wasi --release` and then rerun using the initial `spin up` command whenever changes occur in either Rust (`.rs`) source files or the `Cargo.toml` file. Glob patterns (i.e. the use of the `*` wildcard) are accepted, as shown below:
```

Author: Oooh, I like the "if necessary" part, this allows us to have a single list of monitored items (even though technically component.files will restart not rebuild). I like this.

```toml
[[component]]
[component.build]
command = "cargo build --target wasm32-wasi --release"
watch = ["src/**/*.rs", "Cargo.toml"]
```

>  If you would prefer Spin watch to only rerun the application (without a rebuild) when changes occur, you can use the `--skip-build` option when running the initial `spin watch` command.

Reviewer: ```suggestion
If you would prefer Spin watch to only rerun the application (without a rebuild) when changes occur, you can use the `--skip-build` option when running the `spin watch` command.  In this case, Spin will ignore the `component.build.watch` section, and monitor only the `spin.toml`, `component.source` and `component.files`.
```

Author: Updated.

>  Spin watch can will also monitor any changes to files specified in the `spin.toml`'s `component.files` areas and will rerun the application when changes occur. For example, any changes to `changing-file.txt` in the `my-files` directory will cause Spin watch to rerun the application:

Reviewer: Per comment above, I think this is better combined with the other "things that are watched".  Otherwise a casual reader could easily see "okay it watches spin.toml and component.build.watch, got it" and overlook this paragraph.

>  ```toml

Reviewer: ...and hence suggest merging this with the example above (or keeping two examples but having them both illustrate the unified list).

Author: Done, thanks. Simpler and cleaner.

[[component]]
// -- snip
files = ["my-files/changing-file.txt"]
[component.build]
command = "cargo build --target wasm32-wasi --release"
watch = ["src/**/*.rs", "Cargo.toml"]
```

> Spin watch will respond to filesystem events every 100 milliseconds, by default. This interval can be overridden by passing in the `--debounce` option i.e. `spin watch --debounce 1000` will make Spin watch respond to filesystem events once per second.

Reviewer: ```suggestion
Spin watch waits up to 100 milliseconds before responding to filesystem events, then processes all events that occurred in that interval together. This is so that if you make several changes close together (for example, using a Save All command), you get them all processed in one rebuild/reload cycle, rather than going through a cycle for each one. You can override the interval by passing in the `--debounce` option; e.g. `spin watch --debounce 1000` will make Spin watch respond to filesystem events at most once per second.
```

Author: Done, thanks!

> > Note: Action is only taken after the `--debounce` time has elapsed. For example, when a watched file is changed using a `--debounce` value of 10000 (10 seconds) no effect will be seen in the application, until a full 10 seconds has elapsed; at which point rebuild and/or rerun may occur.

Reviewer: Suggest combining this with the paragraph above - they say very similar things and it took me a moment to grasp the difference.  I've tried to incorporate what this says in my suggested revision.

Author: You adjustment caters for this note. I will remove the note.

> Note: If the build step (`spin build`) fails, `spin up` will not be run.

> Passing the `--clear` flag will clear the screen anytime a rebuild or rerun occurs. Spin watch will not clear the screen between rebuild and rerun as this provides you with an opportunity to see any warnings.

Reviewer: ```suggestion
Passing the `--clear` flag clears the screen anytime a rebuild or rerun occurs. Spin watch does not clear the screen between rebuild and rerun as this provides you with an opportunity to see any warnings.
```

Author: Updated in next push.

> Passing the `--clear` flag will clear the screen anytime a rebuild or rerun occurs. Spin watch will not clear the screen between rebuild and rerun as this provides you with an opportunity to see any warnings.

Reviewer: ```suggestion
Passing the `--clear` flag clears the screen anytime a rebuild or rerun occurs. Spin watch does not clear the screen between rebuild and rerun as this provides you with an opportunity to see any warnings.
```

> > Note: Spin watch will always rebuild every component in your Spin application. It does not discern which individual component changed.

Reviewer: This is not part of the `spin watch` contract and might well change in future - indeed there is an issue asking for this to be addressed.  I'd recommend not saying anything about it either way.

Author: Ok, will remove.

> Spin watch does not consider changes a file's metadata (when it was last modified) as a change

Reviewer: Maybe move this sentence further up to where we are talking about files and reloads?  It feels a bit orphaned here on the other side of the `--clear` flag.  Or drop it altogether?  I'm not sure it's super important.

Author: I moved this up to where we talk about changes; more relevant.

Some trigger types support additional `spin up` flags.  For example, HTTP applications can have a `--listen` flag to specify an address and port to listen on.  See the [HTTP trigger](http-trigger) and [Redis trigger](redis-trigger) pages for more details.

## Monitoring Applications for Changes

Spin's `watch` command rebuilds and restarts Spin applications whenever files change. You can use the `spin watch` [command](https://developer.fermyon.com/common/cli-reference#watch) in place of the `spin build` and `spin up` commands, to build, run and then keep your Spin application running without manual intervention while staying on the latest code and files.

> The `watch` command accepts valid Spin [up](https://developer.fermyon.com/common/cli-reference#up) options and passes them through to `spin up` for you when running/rerunning the Spin application.

> By default, Spin watch monitors:

Reviewer 2: Regarding all the pluralization comments it's maybe easier to just mention off the bat that it is plural/per-component. Or maybe it is already self-evident — your call.

* The application manifest (`spin.toml` file)

> * Any files specified in the `component.build.watch` section of the `spin.toml` file

Reviewer 2: Nit: Technically section should be plural because there could be multiple `component.build.watch` sections.

> * Any files specified in the `component.build.watch` section of the `spin.toml` file

Author: ```suggestion
* Any files specified in the `component.build.watch` sections of the `spin.toml` file
```

> * Any files specified in the `component.files` section of the `spin.toml` file

Reviewer 2: Nit: Technically section should be plural because there could be multiple `component.files` sections.

> * Any files specified in the `component.files` section of the `spin.toml` file

Author: ```suggestion
* Any files specified in the `component.files` sections of the `spin.toml` file
```

> * The files specified in the `component.source` section of the `spin.toml` file

Reviewer 2: Nit: Technically section should be plural because there could be multiple `component.source` sections.

> * The files specified in the `component.source` section of the `spin.toml` file

Author: ```suggestion
* The files specified in the `component.source` sections of the `spin.toml` file
```

If any of these change, Spin will rebuild the application if necessary, then restart the application with the new files.

> > Spin watch does not consider changes to a file's metadata (when it was last modified) as a change.

Reviewer 2: Nit: Permissions are another example of metadata we ignore.

> > Spin watch does not consider changes to a file's metadata (when it was last modified) as a change.

Author: ```suggestion
> Spin watch does not consider changes to a file's metadata (file permissions or when it was last modified) as a change.
```

> The following `spin.toml` configuration (belonging to a Spin `http-rust` application) is configured to ensure that the application is both **rebuilt** (via `cargo build --target wasm32-wasi --release`) and **rerun** whenever changes occur in any Rust source (`.rs`) files, the `Cargo.toml` file or the `spin.toml` file, itself. When changes occur in either the Wasm binary file (`target/wasm32-wasi/release/test.wasm`) or the text file (`my-files/changing-file.txt`) the application is **rerun** using the initial `spin up` command:

Reviewer 2: ```suggestion
The following `spin.toml` configuration (belonging to a Spin `http-rust` application) is configured to ensure that the application is both **rebuilt** (via `cargo build --target wasm32-wasi --release`) and **rerun** whenever changes occur in any Rust source (`.rs`) files, the `Cargo.toml` file or the `spin.toml` file, itself. When changes occur in either the Wasm binary file (`target/wasm32-wasi/release/test.wasm`) or the text file (`my-files/changing-file.txt`) the application is only **rerun** using the initial `spin up` command:
```
